### ISSUE TITLE
FP zero_ptr in no_std contest

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#10017]
### Summary

Not sure if this is a valid usecase :thinking: 

### Lint Name

zero-ptr

### Reproducer

I tried this code:

```rust
#![crate_type = "rlib"]
#![no_std]

#[inline]
pub unsafe fn allocate(_size: usize, _align: usize) -> *mut u8 { 0 as *mut u8 }

```

I saw this happen:

```
    Checking imcrate v0.1.0 (/tmp/imcrate)
warning: failed to automatically apply fixes suggested by rustc to crate `imcrate`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0433]: failed to resolve: use of undeclared crate or module `std`
 --> src/main.rs:5:66
  |
5 | pub unsafe fn allocate(_size: usize, _align: usize) -> *mut u8 { std::ptr::null_mut::<u8>() }
  |                                                                  ^^^ use of undeclared crate or module `std`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0433`.
Original diagnostics will follow.

error: cannot mix `bin` crate type with others

error[E0601]: `main` function not found in crate `imcrate`
 --> src/main.rs:5:80
  |
5 | pub unsafe fn allocate(_size: usize, _align: usize) -> *mut u8 { 0 as *mut u8 }
  |                                                                                ^ consider adding a `main` function to `src/main.rs`

error: `#[panic_handler]` function required, but not found

error: language item required, but not found: `eh_personality`
  |
  = note: this can occur when a binary crate with `#![no_std]` is compiled for a target where `eh_personality` is defined in the standard library
  = help: you may be able to compile for a target that doesn't need `eh_personality`, specify a target with `--target` or in `.cargo/config`

For more information about this error, try `rustc --explain E0601`.
warning: `0 as *mut _` detected
 --> src/main.rs:5:66
  |
5 | pub unsafe fn allocate(_size: usize, _align: usize) -> *mut u8 { 0 as *mut u8 }
  |                                                                  ^^^^^^^^^^^^ help: try: `std::ptr::null_mut::<u8>()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#zero_ptr
  = note: requested on the command line with `-W clippy::zero-ptr`

error: could not compile `imcrate` due to 4 previous errors
```

### Version

```text
rustc 1.67.0-nightly (c97b539e4 2022-11-30)
binary: rustc
commit-hash: c97b539e408ea353f4fde2f9251d598291fec421
commit-date: 2022-11-30
host: x86_64-unknown-linux-gnu
release: 1.67.0-nightly
LLVM version: 15.0.4
```
